### DISCUSSION TITLE
How to figure out which task returned error in try_join!

### DISCUSSION THREAD

**Discussion by Jimmy-Z** [tokio-rs/tokio#5533] [Q&A] [answered]
`try_join!` actually fits my needs quite nice, when everything goes well, I want all tasks to be finished(rules out `select!`), if anything goes wrong, I want all others to be cancelled(rules out `join!`).

But one problem, I can't figure out which task returned that error, and since they're just bubbled up `std::io::Error`, I can't really tell them apart.

So far, the solution I found is an encapsulating `Error` type with a task _identifier_ in it, something like this:
```rust
enum Error {
    TaskA(std::io::Error),
    TaskB(std::io::Error),
    TaskC(std::io::Error),
}

match tokio::try_join!(
    async { task_a.await.map_err(|e| Error::TaskA(e)) },
    async { task_b.await.map_err(|e| Error::TaskB(e)) },
    async { task_c.await.map_err(|e| Error::TaskC(e)) },
) {
    Ok(_) => {...},
    Err(Error::TaskA(e)) => {...},
    Err(Error::TaskB(e)) => {...},
    Err(Error::TaskC(e)) => {...},
};
```

But that looks a little convoluted, so I'm asking if there's a better solution and I'm doing something stupid.

Thanks in advance.

**Comment by Darksonn** [comment, marked as answer]
The macro doesn't provide anything built-in for this.

**Reply by Jimmy-Z** [reply]
Thanks for the answer, and, do you know a better solution to this though?

**Reply by Darksonn** [reply]
If you handle/modify the error inside the branch, then you are in a location where you know what the error is.
